distributed/protocol.py

```diff
@@ -64,35 +62,77 @@
     default_compression = 'lz4'
 
 
+def _extract_big_bytes(x, big, path=()):
+    if type(x) is dict:
+        for k, v in x.items():
+            if isinstance(v, (list, dict)):
+                _extract_big_bytes(v, big, path + (k,))
+            elif type(v) is bytes and len(v) >= 2**31:
+                big[path + (k,)] = v
+                del x[k]
+    elif type(x) is list:
+        for k, v in enumerate(x):
+            if isinstance(v, (list, dict)):
+                _extract_big_bytes(v, big, path + (k,))
+            elif type(v) is bytes and len(v) >= 2**31:
+                big[path + (k,)] = v
+                x[k] = None
+
+
 def dumps(msg):
```

Comment:
Using these messages as examples

```python
from distributed.protocol import dumps, loads

# Unrealistic and intentionally bad
msg = {'a': list('abcdefghijklmnopqrstuvwxyz'),
       'b': dict(zip(range(1000), range(1000))),
       'c': 'foo',
       'd': 'bar',
       'e': [{'a': 1} for i in range(100)]}

# More realistic
msg2 = {'a': 'foo',
        'b': 'bar',
        'c': 1.5,
        'd': ['foo', 'bar']}

s = dumps(msg)
s2 = dumps(msg2)
```
This PR:

```python
In [1]: %timeit dumps(msg)
1000 loops, best of 3: 590 µs per loop

In [2]: %timeit loads(s)
10000 loops, best of 3: 49.9 µs per loop

In [3]: %timeit dumps(msg2)
The slowest run took 6.03 times longer than the fastest. This could mean that an intermediate result is being cached.
100000 loops, best of 3: 6.48 µs per loop

In [4]: %timeit loads(s2)
The slowest run took 9.48 times longer than the fastest. This could mean that an intermediate result is being cached.
1000000 loops, best of 3: 1.16 µs per loop
```

Master:

```python
In [1]: %timeit dumps(msg)
10000 loops, best of 3: 50.2 µs per loop

In [2]: %timeit loads(s)
10000 loops, best of 3: 49.4 µs per loop

In [3]: %timeit dumps(msg2)
The slowest run took 7.68 times longer than the fastest. This could mean that an intermediate result is being cached.
100000 loops, best of 3: 4.04 µs per loop

In [4]: %timeit loads(s2)
The slowest run took 12.41 times longer than the fastest. This could mean that an intermediate result is being cached.
1000000 loops, best of 3: 1.06 µs per loop
```

In summary:
- No overhead on loads for realistic or unrealistic cases
- Dumps is slower. Depending on the depth/breadth of the tree this overhead may be small or not.

I haven't spent any time optimizing this though, I can probably find some small speedups at least.